Add visual formula editor for `pl-symbolic-input`

nritschel opened the pull request:
When `pl-symbolic-input` is used for more complex math questions, the limiting factor is typically not the SymPy backend, but the UX of entering a long formula into a plain HTML input without any formatting or live preview. Most math-specific online learning platforms like WebAssign offer a visual formula editor to provide a better editing experience for students.

This PR is a cleaned up version of a custom fork of `pl-symbolic-input` that was developed over the Summer by @kikis05, with a focus on the needs of Calculus courses. It uses [MathLive](https://mathlive.io) for the formula editor front-end, which is not the most lightweight solution, but turned out to be a fairly good fit for the element with some customizations (mostly to disable more advanced features and to improve compatibility with SymPy). The same library is also used by the calculator prototype (#11426), so ideally, adding one dependency can serve multiple purposes at once.

We have spent a fair amount of time and effort iterating on this element and figuring out the quirks of both MathLive and how it interacts with SymPy. I think this is ready for some broader review/comments beyond the scope of the course. Nonetheless, I am marking this as a draft for now since this is a pretty hefty addition and we did not yet look into some details like if this meets accessibility standards, whether there are any compatibility issues with older/niche browsers, or if the layout/responsive behavior always matches that of the existing input.

nritschel commented:
Note that I removed the aria-label/labelledby attributes for the math-field to make the accessibility tests pass for now, but I am not sure if this is actually the right solution. More generally, the used Mathfield input is supposed to be accessible and screen reader friendly, but we have not thoroughly checked this. 

Depending on how good the accessibility of the input turns out, it might be a good idea to provide some way for students to opt out of the formula editor (e.g., a small button somewhere) and fall back to the standard text input box instead...

reteps reviewed (changes requested):
First pass -- some modularization/unit tests would be helpful.

nritschel commented on docs/elements.md:
```diff
@@ -1123,6 +1123,7 @@ def generate(data):
 | `label`                         | string                  | —                       | A prefix to display before the input box (e.g., `label="$F =$"`).                                                                                                                                                                                                                                                       ...
 | `aria-label`                    | string                  | —                       | An accessible label for the element.                                                                                                                                                                                                                                                                                    ...
 | `display`                       | `"block"` or `"inline"` | `"inline"`              | How to display the input field.                                                                                                                                                                                                                                                                                         ...
+| `formula-editor`                | boolean                 | false                   | Whether the element should provide a visual formula editor for students (recommended when answers are long or contain nested mathematical expressions).                                                                                                                                                                 ...
```
`allow-trig-functions` is an existing attribute that's listed further down in the table, if that's what you mean.

nritschel commented on apps/prairielearn/elements/pl-symbolic-input/pl-symbolic-input.css:
```diff
@@ -3,6 +3,70 @@
   height: auto;
 }
+
+@media (pointer: coarse) {
+  math-field::part(menu-toggle) {
+    display: none;
+  }
+  math-field::part(virtual-keyboard-toggle) {
+    display: none;
+  }
```
MathLive actually provides its keyboard for mobile devices (and the keyboard subsumes the features of the quick menu), but it is not triggered by the UI button. Removing the UI on mobile, where it is redundant and will likely get in the way given the limited screen space, was intentional.

nritschel commented on apps/prairielearn/elements/pl-symbolic-input/pl-symbolic-input.js:
```diff
@@ -0,0 +1,370 @@
+ * @param {HTMLElement} mf The <math-field> element to initialize
+ */
+function setUpSymbolicInputMacros(mf) {
+  const additionalFunctions = mf.getAttribute('custom-functions')
+    ? JSON.parse(mf.getAttribute('custom-functions').replaceAll("'", '"'))
+    : [];
```
This has become redundant with other changes being made to the code.

nritschel commented:
I've done some accessibility testing, and it seems like overall, the navigation and speaking of math work quite well. MathLive has a built-in text to speech generator that hooks up with the OS Voiceover feature, but can also be manually triggered via Ctrl+Alt/Option+Arrow Up. It usually produces better or at least as good results as a plain text field would.

The bad news is that there seems to be an issue with function name macros like `\sin`, which are rendered as `\operatorname{sin}`. Right now, those are skipped during the speech generation, and after some digging and debugging, I am fairly confident that this is a bug in the MathLive code and not just a configuration issue on our end. I'm currently reaching out to the developer via Discord to see if this is something that needs to be added to the "requested fixes" list (that so far has only affected the calculator), or can be worked around on our end...

(I'll post some demo videos once this has been resolved or I at least know more about the status of this.)

nritschel commented on apps/prairielearn/elements/pl-symbolic-input/pl-symbolic-input.py:
```diff
@@ -378,6 +412,75 @@ def parse(element_html: str, data: pl.QuestionData) -> None:
+    constants_class = psu._Constants()
+    functions = (
+        list(psu.STANDARD_OPERATORS)
+        + list(constants_class.functions.keys())
+        + custom_functions
+    )
+    if allow_trig:
+        functions += list(constants_class.trig_functions.keys())
+    multi_char_variables = [v for v in variables if len(v) > 1]
+    names = functions + multi_char_variables
+
+    # Necessary for names that are prefixes of other names, e.g., acos and acosh
+    names.sort(key=len, reverse=True)
+
+    # Step 2: Merge allowed names (e.g., "s i n" -> "sin")
+    for name in names:
+        spaced_name = " ".join(list(name))
+        if spaced_name in sub:
+            sub = sub.replace(spaced_name, name)
+
+    # Step 3: Protect names that contain numbers (e.g., in custom functions)
+    protected_indices = set()
+    for name in names:
+        if re.search(r"\d", name):
+            for match in re.finditer(re.escape(name), sub):
+                start = match.start()
+                end = match.end()
+                protected_indices.update(range(start, end))
+
+    # Step 4: Add spaces between all letter->number pairs that are not custom names
+    i = 1
+    while i < len(sub):
+        if i in protected_indices:
+            i += 1
+            continue
+        if sub[i].isdigit() and sub[i - 1].isalpha():
+            sub = sub[:i] + " " + sub[i:]
+            i += 1  # Skip over the just inserted space
+        i += 1
+    return sub
```
Good catch! Fixed (and added test to ensure it is)

reteps commented on apps/prairielearn/elements/pl-symbolic-input/pl-symbolic-input.js:
```diff
@@ -0,0 +1,353 @@
+ * @param {HTMLElement} mf The <math-field> element to initialize
+ */
+function setUpSymbolicInputMacros(mf) {
+  const additionalFunctions = mf.getAttribute('custom-functions')
+    ? mf.getAttribute('custom-functions').split(',')
+    : [];
```
```suggestion
  const additionalFunctions = mf.getAttribute('custom-functions')?.split(',')
    ?? [];
```

reteps commented on apps/prairielearn/elements/pl-symbolic-input/pl-symbolic-input.js:
```diff
@@ -0,0 +1,353 @@
+  // Otherwise, an accidental function call like "sin" would be converted into "s*i*n", which leads to
+  // a much harder to understand error message (about "s" being an unknown variable) generated by SymPy.
+  const customFunctions = new Set(
+    additionalFunctions.concat([
```
Can you define this list as a separate variable, like `trigFunctions`

reteps commented on apps/prairielearn/elements/pl-symbolic-input/pl-symbolic-input.js:
```diff
@@ -0,0 +1,353 @@
+      'factorial',
+      'ln',
+      'log',
+      'Max',
```
It looks like you have an uppercase version of a couple of these. Why is that needed and what is the criteria for that?

reteps commented on apps/prairielearn/elements/pl-symbolic-input/pl-symbolic-input.js:
```diff
@@ -0,0 +1,353 @@
+  );
+
+  const macros = {};
+  for (const fun of customFunctions) {
```
nit: Use a `.map` here.

reteps commented on apps/prairielearn/elements/pl-symbolic-input/pl-symbolic-input.py:
```diff
@@ -319,7 +344,16 @@ def parse(element_html: str, data: pl.QuestionData) -> None:
 
     # Get submitted answer or return parse_error if it does not exist
-    a_sub = data["submitted_answers"].get(name, None)
+    if formula_editor:
+        a_sub = format_submission_for_sympy(
+            data["submitted_answers"].get(name, None),
+            allow_trig,
+            variables,
+            custom_functions,
+        )
+    else:
+        a_sub = data["submitted_answers"].get(name, None)
+
```
```suggestion
    submitted_answer = data["submitted_answers"].get(name, None)
    if formula_editor:
        a_sub = format_submission_for_sympy(
            submitted_answer,
            allow_trig,
            variables,
            custom_functions,
        )
    else:
        a_sub = submitted_answer

```

reteps reviewed (commented):
initial review. I have not tested yet, just wanted to ask some exploratory questions.

nritschel commented on apps/prairielearn/src/lib/code-caller/code-caller-native.ts:
```diff
@@ -118,7 +118,7 @@ export class CodeCallerNative implements CodeCaller {
     const pythonExecutable = await run(async () => {
       for (const p of options.pythonVenvSearchPaths) {
-        const venvPython = path.resolve(REPOSITORY_ROOT_PATH, path.join(p, 'bin', 'python3.10'));
+        const venvPython = path.resolve(REPOSITORY_ROOT_PATH, path.join(p, 'bin', 'python3.13'));
```
@reteps Was this an intentional change? The CI tests (but not the Python ones) now all fail with errors that seem unrelated to the element, and this seems like the one change (compared to the last version that passed the CI) that could cause this weirdness...

reteps commented on testCourse/questions/symbolicInputEditor/server.py:
```diff
@@ -0,0 +1,34 @@
+    (y, b, a, m) = sympy.symbols("y b a m")
+    x = (y - b) / a
+    z = m * (sympy.cos(a) + sympy.I * sympy.sin(a))
+    c = a + sympy.I * b
```
We don't care about being super clean for server.py code in the test/exampleCourse.

reteps reviewed (commented):
I discussed with @mwest1066 that we think that the example course should be sufficient for testing. To help with this, I made a version of the example course questions with this enabled for everything.

<img width="1204" height="248" alt="CleanShot 2025-10-17 at 16 41 55@2x" src="https://github.com/user-attachments/assets/a8a8a647-2a7e-46ab-b930-a9bbba1cb42d" />

- we probably want input rules for alpha and beta

<img width="1260" height="412" alt="CleanShot 2025-10-17 at 16 42 46@2x" src="https://github.com/user-attachments/assets/4ca23d1e-f3f7-4e6e-b290-b8772b78545a" />

if this option is set, do we want an input rule that changes `log` to `ln`?

Similarly for this option

<img width="1290" height="422" alt="CleanShot 2025-10-17 at 16 43 28@2x" src="https://github.com/user-attachments/assets/a703303e-49ef-466c-b81a-84e4ebb2fff4" />

Overall, since this will be opt in, I think we are pretty close to shipping an MVP. I didn't notice any other issues.

nritschel commented:
I added the greek letter transformations (and also a new virtual keyboard for greek letters). I don't think we should handle the ln/log and i/j switches in the formula editor. In the best cases, this might be a nice to have for students, but in other cases, it could break questions (e.g., ones that use `j` as imaginary unit to free up `i` as a variable name, which we then really don't want to auto-replace with a `j`).

reteps commented on apps/prairielearn/elements/pl-symbolic-input/pl-symbolic-input.js:
```diff
@@ -0,0 +1,595 @@
+    'Theta',
+    'Iota',
+    'Kappa',
+    'Lamda',
```
@nritschel

nritschel commented:
Alright I've added the keyboard switches based on the imaginary unit and log/ln display settings. I tried the three buttons @reteps suggested, but having those all randomly bunched up in the button row is not much better than having no buttons there at all. I added the suggested shortcuts like $x^3$ as hold-to-select options for the more closely related buttons (e.g., for $x^2$), and I think that's good enough UX-wise.

I also fixed some incredibly finnicky issue where some uppercase Greek letters like Alpha have no Latex equivalent (e.g., $\Alpha$) because they look exactly like the modern alphabet equivalent (e.g., A), which made those get treated like said modern alphabet letter in the submission. I've fixed this and also added some extra handling for consecutive Greek letters (e.g., $\alpha\beta$) that weren't parsed correctly before because of the missing space between them. I've done some pretty extensive testing with combos like `alpha, alpha0, Alpha, Alpha0` as allowed variables in the same input box, and it seems like it's all working as expected even in these extreme cases now.

reteps reviewed (approved):
All my concerns have been addressed, except for some minor visual glitches, which should be addressed before this is merged, but I don't consider them extremely important. I tested this on the `symbolicInput` question in the test course, which shows a variety of cases, all appear to be working.

1. The tooltip for the keyboard icon is not visible anymore.
2. The action icons on the alphabetical keyboard aren't centered.
3. Thanks for fixing the overlap issue. I see now that it wasn't the correct fix, and we should instead be preventing the input from going outside that visible region. The current behavior makes it impossible to see the cursor.

See the attached images:

Can't find my cursor:

<img width="1179" height="2556" alt="image" src="https://github.com/user-attachments/assets/47e66c5f-949c-4018-b18b-0ef54aea73da" />

Odd keyboard:

<img width="1179" height="2556" alt="image" src="https://github.com/user-attachments/assets/cbb9b9de-7c9b-4b19-9289-06e8a9ef8a35" />

Bad tooltip:

<img width="1730" height="690" alt="CleanShot 2025-10-23 at 18 33 29@2x" src="https://github.com/user-attachments/assets/d3080d2c-c912-4e53-9af1-6a8d54b9eb25" />

nritschel commented on apps/prairielearn/elements/pl-symbolic-input/pl-symbolic-input.js:
```diff
@@ -0,0 +1,649 @@
+
+    mf.addEventListener('focus', () => {
+      mathVirtualKeyboard.layouts = [elementKeyboardLayout, ...defaultKeyboardLayouts];
+    });
```
If we have a misload then a lot more is going to break than this...

nritschel commented on apps/prairielearn/elements/pl-symbolic-input/pl-symbolic-input.py:
```diff
@@ -319,14 +344,33 @@ def parse(element_html: str, data: pl.QuestionData) -> None:
+            data["format_errors"][name] = "No submitted answer."
+            data["submitted_answers"][name] = None
+            return
+
```
I've added a type guard, even though this issue isn't introduced by the formula editor (and I'm not 100% convinced that we could even encounter this in practice).

reteps commented:
🚢 it!
